File: src/0079.cpp
// warning : TLE code

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        unordered_map<char, int> boardic;
        for(int i = 0; i < board.size(); ++i) {
            for(int j = 0; j < board[0].size(); ++j) {
                char c = board[i][j];
                if(boardic.count(c)) {
                    boardic[c] += 1;
                }else {
                    boardic[c] = 1;
                }
            }
        }

        unordered_map<char, int> wordic;
        for(char c : word) {
            if(wordic.count(c)) {
                wordic[c] += 1;
            }else {
                wordic[c] = 1;
            }
        }

        for(pair<char, int> kv : wordic) {
            if(not boardic.count(kv.first) || boardic[kv.first] < kv.second) {
                return false;
            }
        }

        for(int i = 0; i < board.size(); ++i) {
            for(int j = 0; j < board[0].size(); ++j) {
                if(board[i][j] == word[0]) {
                    if(dfs(board, word, i, j, 0)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    bool dfs(vector<vector<char>>& board, string& word, int i, int j, int k) {
        if(i < 0 || j < 0 || i >= board.size() || j >= board[0].size()) {
            return false;
        }
        if(k >= word.size() || word[k] != board[i][j]) {
            return false;
        }

        if(k == word.size() - 1) {
            return true;
        }

        vector<vector<int>> directions {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        for(vector<int> direct : directions) {
            char temp = board[i][j];
            board[i][j] = '.';
            if(dfs(board, word, i+direct[0], j+direct[1], k+1)) {
                return true;
            }
            board[i][j] = temp;
        }
        return false;
    }
};

File: src/0079.py
from collections import Counter, defaultdict

class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        boardic = defaultdict(int)
        for i in range(len(board)):
            for j in range(len(board[0])):
                boardic[board[i][j]] += 1

        wordic = Counter(word)
        for c in wordic:
            if c not in boardic or boardic[c] < wordic[c]:
                return False

        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0]:
                    if self.dfs(i, j, 0, board, word):
                        return True

        return False

    def dfs(self, i, j, k, board, word):
        if i < 0 or j < 0 or i >= len(board) or j >= len(board[0]) or \
        k >= len(word) or word[k] != board[i][j]:
            return False

        if k == len(word) - 1:
            return True

        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        for x, y in directions:
            tmp = board[i][j]
            board[i][j] = -1
            if self.dfs(i + x, j + y, k + 1, board, word): 
                return True
            board[i][j] = tmp
